Thread safe loader 2

mattp- opened the pull request:
### What does this PR do?
this is a version of #45782 with latest develop merged in. I've not made any changes other than resolving merge conflicts; the original jenkins results have aged out so I'm creating this to re-run. If I figure out a solution I'll send PR to @smarsching and close this.

### What issues does this PR fix or reference?

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

Yes/No

### Commits signed with GPG?

Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

smarsching committed:
Fix race condition in Salt loader.

There was a race condition in the salt loader when injecting global
values (e.g. "__pillar__" or "__salt__") into modules. One effect of
this race condition was that in a setup with multiple threads, some
threads may see pillar data intended for other threads or the pillar
data seen by a thread might even change spuriously.

There have been earlier attempts to fix this problem (#27937, #29397).
These patches tried to fix the problem by storing the dictionary that
keeps the relevant data in a thread-local variable and referencing this
thread-local variable from the variables that are injected into the
modules.

These patches did not fix the problem completely because they only
work when a module is loaded through a single loader instance only.
When there is more than one loader, there is more than one
thread-local variable and the variable injected into a module is
changed to point to another thread-local variable when the module is
loaded again. Thus, the problem resurfaced while working on #39670.

This patch attempts to solve the problem from a slightly different
angle, complementing the earlier patches: The value injected into the
modules now is a proxy that internally uses a thread-local variable to
decide to which object it points. This means that when loading a module
again through a different loader (possibly passing different pillar
data), the data is actually only changed in the thread in which the
loader is used. Other threads are not affected by such a change.

This means that it will work correctly in the current situation where
loaders are possibly created by many different modules and these
modules do not necessary know in which context they are executed. Thus
it is much more flexible and reliable than the more explicit approach
used by the two earlier patches.

Unfortunately, the stand JSON and Msgpack serialization code cannot
handle proxied objects, so they have to be unwrapped before passing them
to that code.

The salt.utils.json module has been modified to  takes care of
unwrapping objects that are proxied using the ThreadLocalProxy.

The salt.utils.msgpack module has been added and basically provides the
same functions as the salt.utils.json module, but for msgpack. Like the
json module, it takes care of unwrapping proxies.

damon-atkins commented:
Go Go Jenkins!

DmitryKuzmenko reviewed (approved):
LGTM

gtmanfred commented on salt/utils/msgpack.py:
```diff
@@ -0,0 +1,85 @@
+
+def pack(o, stream, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Fluorine

gtmanfred commented on salt/utils/msgpack.py:
```diff
@@ -0,0 +1,85 @@
+
+def packb(o, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Fluorine

gtmanfred commented on salt/utils/msgpack.py:
```diff
@@ -0,0 +1,85 @@
+
+def unpack(stream, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Fluorine

gtmanfred commented on salt/utils/msgpack.py:
```diff
@@ -0,0 +1,85 @@
+
+def unpackb(packed, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Fluorine

gtmanfred commented on salt/utils/thread_local_proxy.py:
```diff
@@ -0,0 +1,738 @@
+Proxy object that can reference different values depending on the current
+thread of execution.
+
+..versionadded:: Oxygen
```
Fluorine

gtmanfred reviewed (changes requested):
LGTM, needs the versionadded references updated to Fluorine.

[![blind-review](https://github.com/blindreviews3.png)](https://github.com/zombie/blind-reviews "Review performed in blind mode")

gtmanfred commented:
Also, looks like there are some pylint warnings in salt/state.py https://jenkinsci.saltstack.com/job/pr-lint/job/PR-48598/1/warnings52Result/

cachedout commented on salt/transport/zeromq.py:
```diff
@@ -433,7 +433,14 @@ def __del__(self):
         if not self.auth.authenticated:
             yield self.auth.authenticate()
-        self.publish_port = self.auth.creds['publish_port']
+
```
I'm confused about the relationship between loader thread-safety and these proposed changes to the transports.

mattp- commented:
Sorry, that accidentally snuck in. It's from my other PR that I put on
develop, not a feature branch. You can ignore.

On Mon, Jul 16, 2018, 5:57 PM Mike Place <notifications@github.com> wrote:

> *@cachedout* commented on this pull request.
> ------------------------------
>
> In salt/transport/zeromq.py
> <https://github.com/saltstack/salt/pull/48598#discussion_r202837848>:
>
> > @@ -433,7 +433,14 @@ def __del__(self):
>      def connect(self):
>          if not self.auth.authenticated:
>              yield self.auth.authenticate()
> -        self.publish_port = self.auth.creds['publish_port']
> +
>
> I'm confused about the relationship between loader thread-safety and these
> proposed changes to the transports.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/48598#pullrequestreview-137612747>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACtVkRPslxi3_2VhSmeGf2YrC2kBLr2ks5uHQxLgaJpZM4VQclz>
> .
>

cachedout commented:
@isbm Does anybody on @saltstack/team-suse want to review this? We're getting close to merging this in.

cachedout commented:
@mattp- Can you rebase against the head of develop so we have only the related changes? That should clear out those transport changes so we know exactly what we're dealing with here.

mattp- commented:
@cachedout yep will do shortly. to be clear I haven't made any changes yet; though I did notice one incompatibility between how data.serializers.msgpack and data.utils.msgpack is handled that I need to look at more.

mattp- committed:
Merge remote-tracking branch 'upstream/develop' into thread-safe-loader-2

mattp- commented:
unassociated commit removed

cachedout commented:
@mattp- Cool. We also need the issues raised by @gtmanfred resolved before we can move forward, please.

mattp- committed:
update versionadded docstrings

isbm reviewed (approved):
LGTM

mattp- committed:
revert call to utils.msgpack directly

salt.serializers.msgpack indirectly calls it down the stack, alongside other
backcompat managament.

Commit:
Merge branch 'develop' into thread-safe-loader-2

cachedout commented:
I spent some time looking at this today and I'm honestly not sure what's happening. It's failing in different ways on different platforms and, oddly enough, it seems to work just fine on my development machine.

Commit:
Merge branch 'develop' into thread-safe-loader-2

mattp- committed:
Merge branch 'develop' into thread-safe-loader-2

mattp- committed:
Merge remote-tracking branch 'upstream/develop' into thread-safe-loader-2

mattp- committed:
reduce schedule debug logs to trace due to noisiness

mattp- committed:
Merge branch 'develop' into thread-safe-loader-2

mattp- committed:
Merge branch 'develop' into thread-safe-loader-2

mattp- committed:
Merge branch 'develop' into thread-safe-loader-2

mattp- committed:
Merge branch 'develop' into thread-safe-loader-2

Commit:
Merge branch 'develop' into thread-safe-loader-2

Commit:
Merge branch 'develop' into thread-safe-loader-2

mattp- commented:
closing in favor of #50648
